Email blaster

fchensan opened the pull request:
Email blasting system for admin. 

It has its own table listing subscribers with their names, email addresses, and whether they want to receive marketing email or not.

fchensan committed:
modified:   db/schema.rb
	modified:   db/seeds.rb

fchensan committed:
new file:   app/models/subscriber.rb
	new file:   db/migrate/20190908092258_create_subscribers.rb
	new file:   spec/factories/subscribers.rb
	new file:   spec/models/subscriber_spec.rb

fchensan committed:
new file:   app/assets/javascripts/admin/subscriber.coffee
new file:   app/assets/javascripts/admin/subscribers.coffee
new file:   app/assets/stylesheets/admin/subscribers.scss
new file:   app/controllers/admin/subscribers_controller.rb
new file:   app/helpers/admin/subscriber_helper.rb
new file:   app/helpers/admin/subscribers_helper.rb
modified:   app/views/admin/admin/index.html.erb
modified:   app/views/admin/events/new.html.erb
new file:   app/views/admin/subscribers/confirm_email_sent.html.erb
new file:   app/views/admin/subscribers/index.html.erb
new file:   app/views/admin/subscribers/send_email.html.erb
new file:   app/views/admin/subscribers/write_email.html.erb
modified:   config/initializers/assets.rb
modified:   config/routes.rb
renamed:    db/migrate/20190908092258_create_subscribers.rb -> db/migrate/20190926130108_create_subscribers.rb
modified:   db/schema.rb
modified:   db/seeds/development/01_create_users.rb
new file:   db/seeds/development/07_create_subscribers.rb
new file:   spec/controllers/admin/subscriber_controller_spec.rb
new file:   spec/controllers/admin/subscribers_controller_spec.rb
modified:   spec/factories/subscribers.rb
new file:   spec/helpers/admin/subscriber_helper_spec.rb
new file:   spec/helpers/admin/subscribers_helper_spec.rb

fchensan committed:
modified:   Gemfile.lock
	modified:   app/controllers/admin/subscribers_controller.rb
	modified:   app/mailers/application_mailer.rb
	modified:   app/views/admin/admin/index.html.erb
	modified:   app/views/admin/subscribers/confirm_email_sent.html.erb
	modified:   app/views/admin/subscribers/index.html.erb
	modified:   app/views/admin/subscribers/write_email.html.erb
	new file:   app/views/application_mailer/blast_email.html.erb
	new file:   db/migrate/20190928012001_drop_column_in_subscribers.rb
	modified:   db/schema.rb

fchensan committed:
modified:   app/models/ability.rb
	modified:   app/views/admin/admin/index.html.erb

stevenwjy commented on app/controllers/admin/subscribers_controller.rb:
```diff
@@ -0,0 +1,25 @@
+
+	end
+
+end
```
```suggestion
end

```

stevenwjy replied:
```suggestion
end

```

stevenwjy commented on app/controllers/admin/subscribers_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class Admin::SubscribersController < ApplicationController
+	def index
+		@subscribers = Subscriber.all
```
In this repo, we usually use two spaces for indentation because '\t' or tab character can be translated into different amount of white spaces

stevenwjy commented on app/views/admin/subscribers/index.html.erb:
```diff
@@ -0,0 +1,28 @@
+		<div class="title">
+			Email Blaster
+		</div>
+		<%= button_to("Send New Email", admin_write_email_path, method: "get", class: "btn btn-info") %>
```
Hmm i think this button should only be available to people who have permission to sent email? because you have defined `can [:email], Subscriber` in `app/models/ability.rb`

stevenwjy commented on db/schema.rb:
```diff
@@ -121,7 +129,6 @@
     t.datetime "claimed_at"
     t.string "claimed_by"
     t.bigint "customer_id"
-<<<<<<< HEAD
```
Ups, sorry, I just realised that somehow there's a bug in the schema from previous merge conflict, I will update the development branch and you can rebase from the latest commit.

stevenwjy commented on db/seeds.rb:
```diff
@@ -8,3 +8,6 @@
 
 Dir[File.join(Rails.root, 'db', 'seeds', '*.rb')].sort.each { |seed| load seed }
 Dir[File.join(Rails.root, 'db', 'seeds', Rails.env, '*.rb')].sort.each { |seed| load seed }
+usr = User.create(username:'francisco', password:'123', email:'chensanfrancisco@gmail.com', confirmed_at:DateTime.now)
```
you can just use 'admin' and 'superuser' login details in the development environment

stevenwjy replied:
If you prefer to use your own login details for testing, you can insert it into your own database, but don't push the changes to github

stevenwjy commented on db/seeds/development/01_create_users.rb:
```diff
@@ -2,6 +2,7 @@
   admin = User.create(username: 'admin', email: 'admin@example.com',
   password: '123456', confirmed_at: DateTime.now)
   admin.add_role :admin
+  admin.add_role :superuser
```
Please don't do this, because admin is not supposed to have a superuser role. Besides, I just realised that in the seed for development environment, we haven't added superuser. I will update the dev branch and you can rebase from it afterwards.

stevenwjy commented on db/seeds/development/07_create_subscribers.rb:
```diff
@@ -0,0 +1,2 @@
+Subscriber.create(name: 'Francisco', email:'chensanfrancisco@gmail.com', receive_important_email: true, receive_marketing_email: false)
+Subscriber.create(name: 'Elon Musk', email:'e0407688@u.nus.edu', receive_important_email: true, receive_marketing_email: true)
```
Please don't put your own email here, can just use dummy email because it will only be caught by mailcatcher. Otherwise, you can also just insert it into your own database directly from console.

stevenwjy replied:
Also, in this seed, you still have column `receive_important_email` which you have dropped

stevenwjy commented on app/models/ability.rb:
```diff
@@ -11,6 +11,7 @@ def initialize(user)
         can [:create, :update, :destroy], FrequentlyAskedQuestion
         can [:create, :update, :destroy], Testimony
         can [:create, :update], Ticket
+        can [:email], Subscriber
```
I think it would be better if you just grant this ability to superuser

stevenwjy commented on db/migrate/20190928012001_drop_column_in_subscribers.rb:
```diff
@@ -0,0 +1,5 @@
+class DropColumnInSubscribers < ActiveRecord::Migration[5.2]
```
I think since this is still at your own branch, you can force change the migration file at `db/migrate/20190926130108_create_subscribers.rb` to avoid adding unnecessary files at the dev environment

stevenwjy replied:
So you dont need this migration file to just remove one column.

stevenwjy commented on app/mailers/application_mailer.rb:
```diff
@@ -1,13 +1,44 @@
+
+		subscribers = get_subscribers(params[:is_marketing_email])
+
+		mail(to: generate_email_array(subscribers), 
```
For this part, I think it would be better if you can personalise the email by sending it to each receiver one by one, because if you put all emails in the target receiver then everyone will know that you broadcast the emails and get info about other people who receive the email.

stevenwjy replied:
Also, in the email content, you can also add link to unsubscribe and create another controller for it.

stevenwjy commented on db/schema.rb:
```diff
@@ -108,6 +108,14 @@
 
+  create_table "subscribers", force: :cascade do |t|
+    t.string "name"
+    t.string "email"
```
I think you need to make sure that each email can be associated with at most one subscriber by putting unique constraint on email. You can write the validation at `app/models/subscriber.rb`.

stevenwjy commented on app/views/admin/subscribers/send_email.html.erb:
```diff
@@ -0,0 +1,4 @@
+<section class="container">
```
What is this for? The API for send_email is POST request right, so i don't think you need to have this page. Instead, you might want to give another kind of notification if a user fails to send the email (e.g. because of internal error).

fchensan committed:
Update app/controllers/admin/subscribers_controller.rb

Co-Authored-By: Steven Wijaya <stevenwjy149@gmail.com>

fchensan committed:
modified:   app/controllers/admin/subscribers_controller.rb
	modified:   app/mailers/application_mailer.rb
	modified:   app/models/ability.rb
	modified:   app/models/subscriber.rb
	modified:   app/views/admin/subscribers/confirm_email_sent.html.erb
	modified:   app/views/admin/subscribers/index.html.erb
	deleted:    app/views/admin/subscribers/send_email.html.erb
	modified:   app/views/application_mailer/blast_email.html.erb
	modified:   config/routes.rb
	modified:   db/migrate/20190926130108_create_subscribers.rb
	deleted:    db/migrate/20190928012001_drop_column_in_subscribers.rb
	modified:   db/seeds.rb
	modified:   db/seeds/development/01_create_users.rb
	deleted:    db/seeds/development/07_create_subscribers.rb

fchensan committed:
modified:   app/controllers/admin/subscribers_controller.rb
	modified:   app/mailers/application_mailer.rb
	modified:   app/models/subscriber.rb
	modified:   app/views/admin/subscribers/index.html.erb
	deleted:    app/views/application_mailer/blast_email.html.erb
	new file:   app/views/application_mailer/mail_personal.html.erb
	modified:   config/routes.rb
	modified:   db/migrate/20190902174802_create_customers.rb
	modified:   db/migrate/20190926130108_create_subscribers.rb
	modified:   db/schema.rb

fchensan committed:
modified:   app/controllers/admin/subscribers_controller.rb
	modified:   app/mailers/application_mailer.rb
	modified:   app/models/subscriber.rb
	modified:   app/views/admin/events/new.html.erb
	modified:   app/views/admin/subscribers/confirm_email_sent.html.erb
	modified:   app/views/admin/subscribers/index.html.erb
	modified:   app/views/admin/subscribers/write_email.html.erb
	modified:   config/routes.rb
	modified:   db/migrate/20190902174802_create_customers.rb
	modified:   db/migrate/20190926130108_create_subscribers.rb

fchensan committed:
global

fchensan committed:
Added sync_with_customer to SubscribersController

fchensan committed:
Correct column order

fchensan committed:
Add after_create on Customer to create subscriber

fchensan committed:
Set receive_marketing_email in Subscriber default to true

fchensan committed:
fix schema

fchensan committed:
Add sync with customer button

fchensan committed:
Add template to email blaster view

fchensan committed:
Fix unsubscribe behavior

fchensan committed:
Fix unsubscribe

fchensan committed:
Fix unsubscribe

fchensan committed:
Fix problem of html tags not showing up in email

fchensan committed:
Change migration version

fchensan committed:
Ensure superuser in SubscriberController

fchensan committed:
Add email blaster preview

fchensan committed:
Fix blast email template

fchensan committed:
Add greeting field in write email

fchensan committed:
Fix columns

fchensan committed:
Change to deliver_now

fchensan committed:
Add rescue if unsubscribe code not found

stevenwjy commented on db/seeds.rb:
```diff
@@ -7,4 +7,4 @@
 
 Dir[File.join(Rails.root, 'db', 'seeds', '*.rb')].sort.each { |seed| load seed }
-Dir[File.join(Rails.root, 'db', 'seeds', Rails.env, '*.rb')].sort.each { |seed| load seed }
+Dir[File.join(Rails.root, 'db', 'seeds', Rails.env, '*.rb')].sort.each { |seed| load seed }
```
```suggestion
Dir[File.join(Rails.root, 'db', 'seeds', Rails.env, '*.rb')].sort.each { |seed| load seed }

```

stevenwjy committed:
Update db/seeds.rb

fchensan committed:
Merge branch 'email-blaster' of https://github.com/nuansa-cp/gapura into email-blaster

stevenwjy committed:
Merge branch 'development' into email-blaster
